lib/service/modules/query.js

```diff
@@ -0,0 +1,900 @@
+   * descending.
+   *
+   * @param {string} field Field
+   * @returns {Query} The query.
+   */
+  descending(field) {
+    if (this._parent) {
+      this._parent.descending(field);
+    } else {
+      this.sort[field] = -1;
+    }
+
+    return this;
+  }
+
+  /**
+   * Adds a filter to the query.
+   *
+   * @param {string} field Field
+   * @param {string} condition Condition
+   * @param {*} values Values
+   * @returns {Query} The query.
+   */
+  addFilter(field, condition, values) {
+    if (typeof this.filter[field] !== 'object') {
+      this.filter[field] = {};
+    }
+
+    if (condition && values) {
+      this.filter[field][condition] = values;
+    } else {
+      this.filter[field] = condition;
+    }
+
+    return this;
+  }
+
+  /**
+   * @private
+   * Joins the current query with another query using an operator.
+   *
+   * @param {string} operator Operator
+   * @param {Query[]|Object[]} queries Queries
+   * @throws {QueryError} `query` must be of type `Query[]` or `Object[]`.
+   * @returns {Query} The query.
+   */
+  join(operator, queries) {
+    let that = this;
+    const currentQuery = {};
+
+    // Cast, validate, and parse arguments. If `queries` are supplied, obtain
+    // the `filter` for joining. The eventual return function will be the
+    // current query.
+    queries = queries.map(query => {
+      if (!(query instanceof Query)) {
+        if (typeof query === 'object') {
+          query = new Query(query);
+        } else {
+          throw new QueryError('query argument must be of type: Kinvey.Query[] or Object[].');
+        }
+      }
+
+      return query.toJSON().filter;
+    });
+
+    // If there are no `queries` supplied, create a new (empty) `Query`.
+    // This query is the right-hand side of the join expression, and will be
+    // returned to allow for a fluent interface.
+    if (queries.length === 0) {
+      that = new Query();
+      queries = [that.toJSON().filter];
+      that.parent = this; // Required for operator precedence and `toJSON`.
+    }
+
+    // Join operators operate on the top-level of `filter`. Since the `toJSON`
+    // magic requires `filter` to be passed by reference, we cannot simply re-
+    // assign `filter`. Instead, empty it without losing the reference.
+    const members = Object.keys(this.filter);
+    forEach(members, member => {
+      currentQuery[member] = this.filter[member];
+      delete this.filter[member];
+    });
+
+    // `currentQuery` is the left-hand side query. Join with `queries`.
+    this.filter[operator] = [currentQuery].concat(queries);
+
+    // Return the current query if there are `queries`, and the new (empty)
+    // `PrivateQuery` otherwise.
+    return that;
+  }
+
+  /**
+   * @private
+   * Processes the data by applying fields, sort, limit, and skip.
+   *
+   * @param {Array} data The raw data.
+   * @throws {QueryError} `data` must be of type `Array`.
+   * @returns {Array} The processed data.
+   */
+  process(data) {
```

Comment:
`process` is also one of the nodejs globals (process, console, global, module, exports, Buffer, require, __filename, __dirname and setImmediate et al)